File: 2023/day02/day02.go

    package main

    /*
    Day 02: https://adventofcode.com/2023/day/2
    */

    import (
        "flag"
        "fmt"
        "io/ioutil"
        "strings"
        "regexp"
        "strconv"
    )
    
    var inputFile = flag.String("input", "../inputs/day02.txt", "Relative path to input-file")
    
    func star1(ins []string) {
        ans := 0
        for i, ln := range ins {
            s1 := strings.Split(ln, ":")
            s2 := strings.Split(s1[1], ";")
            check := true
            for _,s3 := range s2 {
                r := regexp.MustCompile(`(\d+ (red|green|blue))`)
                matches := r.FindAllString(s3, -1)
                cube := make(map[string]int)
                cube["blue"] = 0
                cube["green"] = 0
                cube["red"] = 0
                for _, v := range matches {
                    tmp := strings.Split(v, " ")
                    cube[tmp[1]],_ = strconv.Atoi(tmp[0])
                    if (cube["red"] >12 || cube["green"] >13 || cube["blue"] >14) {
                        check = false
                        break
                    }
                }   
            }
            if (check) {
                ans += i+1
            }
        }
        fmt.Println("Star 1:", ans)
    }

    func star2(ins []string) {
        ans := 0
        for _, ln := range ins {
            s1 := strings.Split(ln, ":")
            s2 := strings.Split(s1[1], ";")
            power := 0
            min_red := 0
            min_blue := 0
            min_green := 0
            for _,s3 := range s2 {
                r := regexp.MustCompile(`(\d+ (red|green|blue))`)
                matches := r.FindAllString(s3, -1)
                cube := make(map[string]int)
                cube["blue"] = 0
                cube["green"] = 0
                cube["red"] = 0
                
                for _, v := range matches {
                    tmp := strings.Split(v, " ")
                    cube[tmp[1]],_ = strconv.Atoi(tmp[0])
                    if (cube["red"] > min_red) {
                        min_red = cube["red"]
                    }
                    if (cube["green"] > min_green) {
                        min_green = cube["green"]
                    }
                    if (cube["blue"] > min_blue) {
                        min_blue = cube["blue"]
                    }
                } 
            }
            power = min_blue*min_green*min_red 
            fmt.Println(power)
            ans += power
        }
        fmt.Println("Star 2:", ans)
    }

    func main() {
        flag.Parse()
        raw, err := ioutil.ReadFile(*inputFile)
        if err != nil {
            panic("Invalid file")
        }
        rawStr := strings.TrimSpace(string(raw))
        data := strings.Split(rawStr, "\n")
        star1(data)
        star2(data)
    }
    

File: 2025/day02.py
"""
Day 02

https://adventofcode.com/2025/day/2
"""

from typing import List, Tuple, Dict, Set, Optional
from pathlib import Path
import textwrap


def read_input(file_path: str = "./inputs/day02.in") -> List[str]:
    f = Path(file_path).read_text().strip().split("\n")
    return f


def star1(input):
    ans = 0
    for pair in input[0].split(','):
        p1,p2 = pair.split('-')
        i = int(p1)
        while i <= int(p2):
            s = str(i)
            a,b = s[:len(s)//2], s[len(s)//2:]
            if a == b:
                #print(p1, p2, s)
                ans += int(s)
            i+=1
    return ans

def isInvalid(i):
    s = str(i)
    if len(set(s)) > len(s)//2:
        return False
    for n in range(len(s)//2):
        parts = textwrap.wrap(s,n+1)
        if len(set(parts))==1:
            return True

    return False

def star2(input):
    ans = 0
    for pair in input[0].split(','):
        p1,p2 = pair.split('-')
        i = int(p1)
        while i <= int(p2):
            if isInvalid(i):
                ans += int(i)
            i+=1
    return ans
    return 0


def main():
    input = read_input()
    ans1 = star1(input)
    ans2 = star2(input)
    print("Star 1:", ans1)
    print("Star 2:", ans2)


if __name__ == "__main__":
    main()
